thread.c

```diff
@@ -581,6 +581,74 @@ static void thread_libevent_process(evutil_socket_t fd, short which, void *arg)
 /* Which thread we assigned a connection to most recently. */
 static int last_thread = -1;
 
+/* Last thread we assigned to a connection based on napi_id */
+static int last_thread_by_napi_id = -1;
+
+static LIBEVENT_THREAD *select_thread_round_robin(void)
+{
+    int tid = (last_thread + 1) % settings.num_threads;
+
+    last_thread = tid;
+
+    return threads + tid;
+}
+
+static void reset_threads_napi_id(void)
+{
+    LIBEVENT_THREAD *thread;
+    int i;
+
+    for (i = 0; i < settings.num_threads; i++) {
+         thread = threads + i;
+         thread->napi_id = 0;
+    }
+
+    last_thread_by_napi_id = -1;
+}
+
+/* Select a worker thread based on the NAPI ID of an incoming connection
+ * request. NAPI ID is a globally unique ID that identifies a NIC RX queue
+ * on which a flow is received.
+ */
+static LIBEVENT_THREAD *select_thread_by_napi_id(int sfd)
+{
+    LIBEVENT_THREAD *thread;
+    int napi_id, err, i;
+    socklen_t len;
+    int tid = -1;
+
+    len = sizeof(socklen_t);
+    err = getsockopt(sfd, SOL_SOCKET, SO_INCOMING_NAPI_ID, &napi_id, &len);
+    if ((err == -1) || (napi_id == 0))
+        return select_thread_round_robin();
+
+select:
+    for (i = 0; i < settings.num_threads; i++) {
+         thread = threads + i;
+         if (last_thread_by_napi_id < i) {
+             thread->napi_id = napi_id;
+             last_thread_by_napi_id = i;
+             tid = i;
+             break;
+         }
+         if (thread->napi_id == napi_id) {
+             tid = i;
+             break;
+         }
+    }
+
+    if (tid == -1) {
+        fprintf(stderr, "Connection received with unexpected NAPI ID. Resetting all napi ids associated with worker threads. You may need to restart the application if you are noticing any performance drops.\n");
```

Comment:
been trying to not add more stderr stuff. there's the LOGGER_LOG bits (which are super complex, sorry...) and stats counters which people do tend to monitor. ie; record the number of times it gets reset.

what situation would cause performance drops?

Reply:
(er, record the number of times it gets reset and document what that means in doc/protocol.txt)

Reply:
This condition is documented towards the end of napi_ids.txt
    During a normal run, each worker thread gets associated with a napi_id and
    this will establish a 1:1 mapping between the thread and queues. If a new
    napi_id is received after each thread is associated with its own napi_id
    (this can happen if num_napi_ids argument doesn't exactly match with the
    number of RX queues OR if the NIC driver goes through a reload), an error
    message is logged and all the napi_id's associated with the threads are
    reset.

Are you suggesting removing the fprintf() and add a counter to record the number of times reset is called? Can you give an example of such counter that is documented in doc/protocol.txt?

Reply:
listen_disabled_num (poorly named, but is what it is..), oom counters, etc. there're a few of them.

I read the doc about things being reset, but "being reset" doesn't describe the impact to the running service? halve performance, errors, etc?

Reply:
The performance impact would be dependent on the real number of HW queues and actual number of napi_ids, worker threads specified. I don't think i can quantify the impact. Basically this is a misconfiguration and incorrect usage of -N option.

Reply:
Ok, well just to repeat my _intention_ here: managing expectations for end users. Whatever you can think of to help with that is great.